File: C4_DUPLICATES_SR/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

type Node struct {
	key   rune
	count int
	left  *Node
	right *Node
}

func insert(root **Node, key rune) {
	if *root == nil {
		*root = &Node{key: key, count: 1}
	} else if key == (*root).key {
		(*root).count++
	} else if key < (*root).key {
		insert(&(*root).left, key)
	} else {
		insert(&(*root).right, key)
	}
}

func countSingleChildNodes(root *Node) int {
	if root == nil {
		return 0
	}
	count := 0
	if (root.left != nil && root.right == nil) || (root.left == nil && root.right != nil) {
		count = root.count
	}
	return count + countSingleChildNodes(root.left) + countSingleChildNodes(root.right)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	for i := 0; i < 5; i++ {
		input, _ := reader.ReadString('\n')
		input = strings.ToUpper(input)
		var root *Node
		for _, c := range input {
			if unicode.IsLetter(c) {
				insert(&root, c)
			}
		}

		count := countSingleChildNodes(root)
		fmt.Println(count)
	}
}

File: C4_DUPLICATES_SR/Main.java
import java.util.Scanner;

class Node {
  char key;
  int count;
  Node left, right;

  Node(char key) {
    this.key = key;
    this.count = 1;
    this.left = null;
    this.right = null;
  }
}

public class Main {
  static Node insert(Node root, char key) {
    if (root == null) {
      root = new Node(key);
    } else if (key == root.key) {
      root.count++;
    } else if (key < root.key) {
      root.left = insert(root.left, key);
    } else {
      root.right = insert(root.right, key);
    }
    return root;
  }

  static int countSingleChildNodes(Node root) {
    if (root == null) {
      return 0;
    }
    int count = 0;
    if ((root.left != null && root.right == null) || (root.left == null && root.right != null)) {
      count = root.count;
    }
    return count + countSingleChildNodes(root.left) + countSingleChildNodes(root.right);
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    for (int i = 0; i < 5; i++) {
      String input = scanner.nextLine().toUpperCase();
      Node root = null;
      for (int j = 0; j < input.length(); j++) {
        char c = input.charAt(j);
        if (Character.isLetter(c)) {
          root = insert(root, c);
        }
      }

      int count = countSingleChildNodes(root);
      System.out.println(count);
    }
    scanner.close();
  }
}